assets/js/workflow-reducer.js

```diff
@@ -25,312 +41,567 @@ export function mockAPI(mock_api) {
-      var selectedWfModule = null;
-      if (action.workflow.selected_wf_module) {
-        selectedWfModule = action.workflow.selected_wf_module;
-      } else if (action.workflow.wf_modules && action.workflow.wf_modules.length) {
-        selectedWfModule = action.workflow.wf_modules[0].id;
-      };
-      return Object.assign({}, state, {
-        selected_wf_module: selectedWfModule,
-        workflow: action.workflow
-      });
-
-
-    // Change id of module currently selected
-    case SELECTED_MODULE_CHANGE:
-      // console.log(SELECTED_MODULE_CHANGE + " old " +  state.selected_wf_module + " new " + action.id);
-      if (!'selected_wf_module' in state || (action.id != state.selected_wf_module)) {
-        // call API to set selected WF module in backend
-        api.setSelectedWfModule(state.workflow.id, action.id);
-        return Object.assign({}, state, {
-          selected_wf_module: action.id,
-        });
-      } else {
-        return state;
+  if (credentialIndex >= 0) {
+    return update(state, {
+      loggedInUser: {
+        google_credentials: { $splice: [[credentialIndex, 1]] }
       }
+    });
+  }
 
-    // Change status on a single module
-    case MODULE_STATUS_CHANGE:
-      //console.log(WF_MODULE_STATUS_CHANGE + " " + action.id);
-      if ('wf_modules' in state.workflow) {
-
-        var newState = state;
-
-        for (var wfm of newState.workflow.wf_modules) {
+  return state;
+});
 
-          // Find matching module, and change state if status changed or error message changed
-          if (wfm.id == action.id &&
-              ((wfm.status != action.status) || (wfm.status=='error' && wfm.error_msg != action.error_msg))) {
-            // console.log("actually changed status for " + wfm.id);
+// -- Workflow Module --
 
-            // Create a copy of the wf_module with new status
-            var newWfm = Object.assign({}, wfm, { status: action.status, error_msg: action.error_msg });
+// UPDATE_WF_MODULE
+// Patch a workflow module with new data
 
-            // Clone the state, switch out this one wfm (keep position in wf_modules array )
-            newState = Object.assign({}, state);
-            newState.workflow = Object.assign({}, state.workflow);
-            newState.workflow.wf_modules = state.workflow.wf_modules.map(el => el!==wfm ? el : newWfm);
-          }
+// TODO: We don't validate which fields or types are on
+// a WfModule here. The backend will reject nonexistent
+// fields, but should we do something on the frontend?
+export function updateWfModuleAction(id, data) {
+  return {
+    type: UPDATE_WF_MODULE,
+    payload: {
+      promise: api.updateWfModule( id, data ),
+      data: {
+        id,
+        data
+      }
+    }
+  };
+}
+registerReducerFunc(UPDATE_WF_MODULE + '_PENDING', (state, action) => {
+  let moduleIdx = findIdxByProp(
+    state.workflow.wf_modules,
+    'id',
+    action.payload.id
+  );
+
+  if (typeof moduleIdx !== 'undefined') {
```

Comment:
Who would call with undefined?

Reply:
There's some weird cases where this could happen -- let's say a module is in some kind of queue to get updated (say, if the upstream data changed), and each update takes a long time, and before we get to this module, it gets deleted.